### ISSUE TITLE
Creating a new backend for Vispy but out-of-tree

### ISSUE THREAD

**Issue by ProfFan** [vispy/vispy#2161]
Hi,

First thanks for this great library. I wonder if it is possible for Vispy to load a backend with a reference to a Class instead of a name?

I want to make a new backend, but I don't want to fork off vispy.

Thanks!

**Comment by djhoese** [comment]
Great question! If yoy check out the roadmap on the vispy.org site (https://vispy.org/roadmap.html#opengl-vulkan-and-webgpu), you'll find various things about us wanting to change the low-level "stuff" in vispy. You'll also find this exact request/discussion in this issue: https://github.com/vispy/vispy/issues/1995

The hard part for what you're suggesting is that:

1. A backend is not just a single class. There is at least an Application base class and a Canvas base class. There are also variables related to special metadata like what version of the backend library is being used, etc.
2. The infrastructure for vispy doesn't support this yet.

If you'd like to help with restructuring vispy to support this I'm all for it and would accept any help.

**Comment by ProfFan** [comment]
Hi @djhoese ,

Thanks for the detailed reply! Upon evaluating the code I think the only thing that actually prevents hot-loading of a backend is this:
```python
      for key in backends_to_try:
            name, module_name, native_module_name = BACKENDMAP[key]
            TRIED_BACKENDS.append(name)
            mod_name = 'backends.' + module_name
            __import__(mod_name, globals(), level=1)
            mod = getattr(backends, module_name)
            if not mod.available:
```
which forces the backend to be in vispy because of the hard coded `mod_name`.

Do you think this is correct?

**Comment by djhoese** [comment]
Yes, I think you're right. That's at least the first major hurdle. To be clear, here is a direct link to that chunk of code:

https://github.com/vispy/vispy/blob/a58721afe5d554be67ead1543f872325ed2b4d69/vispy/app/application.py#L232-L238

We can see another main usage of the application is the Canvas object here (the Application is also created in the `__init__`:

https://github.com/vispy/vispy/blob/a58721afe5d554be67ead1543f872325ed2b4d69/vispy/app/canvas.py#L231

I think in the long run I would prefer a "registry" type system where other libraries can declare (either as entry_points in setup.py or by calling a function) a new backend. This registry would essentially replace this `BACKENDMAP` dictionary used above. I think passing an "instance" of a backend to the above `_use` method would work just as well.

I'm not sure it is the best interface, but maybe a backend could be defined as a class with specific properties and methods for accessing its features. This would make it easier to enforce a default set of behaviors and define a minimum interface required to implement a backend. From what I can see it would have to define these properties:

* `capability` : dictionary: Dictionary of possible keyword arguments and their default values.
   https://github.com/vispy/vispy/blob/a58721afe5d554be67ead1543f872325ed2b4d69/vispy/app/base.py#L83-L96
* `available` : bool: Whether or not this backend can be used. This will usually be determined by if the underlying low-level libraries are available (ex. `pyqt5` for the `pyqt5` backend)
* `testable` : bool: Should be included when running tests
* `why_not` : str: Why can't this backend be used (what error/exception was raised when trying to initialize it).
* `which` : tuple: Extra metadata related to the version of the backend and its dependencies. Includes a human readable name for the backend. For example:
   
   ```python
   which = ('PyQt5', QtCore.PYQT_VERSION_STR, QtCore.QT_VERSION_STR)
   ```
   
   This is appended to the information in `vispy.sys_info()` when it lists the backends. This would also need to be updated if we wanted to advertise newly "registered" backends to the user: https://github.com/vispy/vispy/blob/a58721afe5d554be67ead1543f872325ed2b4d69/vispy/util/config.py#L421-L426

* `ApplicationBackend` : class: Backend class
* `CanvasBackend` : class: Backend Canvas class
* `TimerBackend`: class: Backend Timer implementation
* `name`: str: Name of the backend that the user can use to request to use this backend. I'm not sure if this should be on the class which would be less flexible (can't be changed after the class is instantiated) or if it is told to the backend class ("you are named this and this is how I will refer to you").

I think that's everything. I'm sure parts of vispy that access `backend_module` or whatever would have to be updated to point to these properties but this is a start. These properties don't have to be named this either, this is just what is being used right now or makes the most sense based on how it is used (ex. `name` isn't defined in the backend module but each backend has a name defined for it).

**Comment by almarklein** [comment]
Side note: I find the `capability` dict quite confusing. It looks like its use is very limited and I wonder if it could be factored out if we would be refactoring backends anyway...
